### PR TITLE
Fixes for runtime4 for the 5.2 API

### PR DESCRIPTION
**Pull request by mshinwell**
This is currently based on #2971, only the latest changeset should be reviewed here.

The major additions are float16 bigarray support and blake2 digests; see upstream `5.2.0`.

### CODE REVIEW

**Comment by mshinwell** [comment]
I've split that single changeset out to make review easier, as follows:
fed43cff6a858c9173ae92d7dc4508d4f796e713 Remainder of minor changes from PR12599 upstream + another fix for a preproc macro which is not defined any more
bcb2c82342316d5157a32900acd168c07b2b99c9 Fix lack of preproc defn in ocaml/runtime4/interp.c
5c91e33aca5bdbdce89180e4a749d7fb34af4d02 Add blake2 to ocaml/runtime4/gen_primitives.sh
d0c7e88f1a2397ba21fafb169eb5ac69ec28dbe3 Fix lack of preproc defn in ocaml/runtime4/eventlog.c
7038271b5c73a681d99968f25975a51d0802da52 Fix lack of preproc defn in ocaml/runtime4/alloc.c
18340400bb541bc5f66a201f14afc09eaf0c94e0 Fix lack of preprocessor definition in 5.2.0 + add dummy stub
1894de7d07ec478ca25891d82b99b4cbc817126c Add blake2 to runtime4/dune
9e74c20fc7497a38c8c9893a7bc4a92416ccf909 Fix bad merge on ocaml/runtime4/caml/compact.h
bfbdb85c95f7f3dd62100f85b71f8d6b8f5b1b7b Add blake2.h to caml/dune
574ac459b957e6941bc51b5b231eff46b922e103 Port PR12599 from upstream, dynlink.c portion
5e1976abb146d9f8e9c1c1ff76386d65ac68353a Add blake2 to runtime4/Makefile
1a3ef96e0ed1fbe9fd7fae4db37c11faf7f6321c Add caml_xdg_defaults, copied from runtime5, required by 5.2.0
2807e5bc696b1172a31c033e14fe239c3c43ff59 Add missing io.c functions from runtime5
20d2fa93e18f050fe44af893384352217cb405c1 CAMLnoret etc
7e0207aca6185b258280bf77a59d23ec0bcf858e Add blake2 files from upstream
3e0640110b959269b7d119a002cb05ba3b9650f0 Backport PR12815 (Correctly format multi-line locations in exception backtraces, in the style that the compiler driver uses.) to runtime4: upstream 067115a0b032c0d0c6dc5daf1c017cea2be3884f
4a13bdab9dff5b2d6e34f584ec49a7d70db02d3d Port the runtime part of PR12430 to runtime4/
a2453e10cb763dc9309ab19abd2f7dbff86bba4b Backport PR12499 (bigarray: fix deserialization of float16 dynarrays) to runtime4: upstream f0223c9f917929b4860b8b449f50441927743136
7b1f0e8765a1142586466729b1e69b3e1a078d47 Backport PR12357 (Move constructor Bigarray.Float16 last in type Bigarray.kind) to runtime4: upstream daa91dd7ec5e454567cabc63fda43ec3b1cec0f4
30958b89211741c0b390eb356265c6e5c38fc1b3 Backport PR10775 (Support 16-bit floating-point numbers as elements of bigarrays) to runtime4: upstream 08276afddb09dfe2d771bf1a70a2222165e9f6fb

File: ocaml/runtime4/interp.c [hunk 1]
```diff
@@ -1231,7 +1231,7 @@ value caml_interprete(code_t prog, asize_t prog_size)
 
 #ifndef THREADED_CODE
     default:
-#if _MSC_VER >= 1200
+#if defined(_MSC_VER) && _MSC_VER >= 1200
```
Comment (xclerc) [inline ocaml/runtime4/interp.c:1234]: We should probably have the same in `ocaml/runtime`.

File: ocaml/runtime4/caml/prims.h [hunk 2]
```diff
@@ -26,9 +26,7 @@ extern c_primitive caml_builtin_cprim[];
 extern char * caml_names_of_builtin_cprim[];
 
 extern struct ext_table caml_prim_table;
-#ifdef DEBUG
```
Comment (xclerc) [inline ocaml/runtime4/caml/prims.h:29]: We should probably have the same in `ocaml/runtime`.

File: ocaml/runtime4/alloc.c [hunk 3]
```diff
@@ -67,7 +67,7 @@ CAMLexport value caml_alloc (mlsize_t wosize, tag_t tag) {
   return caml_alloc_with_reserved (wosize, tag, 0);
 }
 
-#if NATIVE_CODE
+#if defined(NATIVE_CODE) && NATIVE_CODE
```
Comment (xclerc) [inline ocaml/runtime4/alloc.c:70]: (There are occurrences of `#ifdef NATIVE_CODE` under `ocaml/runtime`.)

File: ocaml/runtime4/caml/compact.h [hunk 4]
```diff
@@ -2,9 +2,9 @@
 /*                                                                        */
```
Comment (xclerc) [inline ocaml/runtime4/caml/compact.h:2]: (The diff of this file is unexpected...)

File: ocaml/runtime4/startup_byt.c [hunk 5]
```diff
@@ -399,7 +399,7 @@ static void do_print_config(void)
          "false");
 #endif
   printf("windows_unicode: %s\n",
-#if WINDOWS_UNICODE
+#if defined(WINDOWS_UNICODE) && WINDOWS_UNICODE
```
Comment (xclerc) [inline ocaml/runtime4/startup_byt.c:402]: (The condition under `ocaml/runtime` is
slightly different, but -I think- equivalent.)

**Review by xclerc — APPROVED** [review]
(I have reviewed `30958b8..fed43cf`.)

File: ocaml/runtime4/interp.c (hunk 1 above)
Comment (mshinwell) [reply ocaml/runtime4/interp.c:1234]: It's just `#ifdef MSC_VER` in 5.2.0 so I don't think we need any runtime5 fix here.  I've harmonized runtime4 with this though.

File: ocaml/runtime4/caml/prims.h (hunk 2 above)
Comment (mshinwell) [reply ocaml/runtime4/caml/prims.h:29]: Actually I'm not sure this change is required, let's try without.

File: ocaml/runtime4/alloc.c (hunk 3 above)
Comment (mshinwell) [reply ocaml/runtime4/alloc.c:70]: (I misread this comment.)  Just using `#ifdef`.

File: ocaml/runtime4/caml/compact.h (hunk 4 above)
Comment (mshinwell) [reply ocaml/runtime4/caml/compact.h:2]: This must have been a bad automatic merge.  This file now matches `flambda-backend/main`.

File: ocaml/runtime4/startup_byt.c (hunk 5 above)
Comment (mshinwell) [reply ocaml/runtime4/startup_byt.c]: Fixed

**Pull request merged**